library/cwm/src/lib/cwm/widget.rb

```diff
@@ -0,0 +1,450 @@
+      if !respond_to?(:widget_type)
+        raise "Widget '#{self.class}' does set its widget type"
+      end
+
+      res = {}
+
+      if respond_to?(:help)
+        res["help"] = help
+      else
+        res["no_help"] = ""
+      end
+      res["label"] = label if respond_to?(:label)
+      res["opt"] = opt if respond_to?(:opt)
+      if respond_to?(:validate)
+        res["validate_function"] = validate_method
+        res["validate_type"] = :function
+      end
+      res["init"] = init_method if respond_to?(:init)
+      res["handle"] = handle_method if respond_to?(:handle)
+      res["store"] = store_method if respond_to?(:store)
+      res["cleanup"] = cleanup_method if respond_to?(:cleanup)
+      res["widget"] = widget_type
+
+      res
+    end
+
+    # gets if widget is open for modification
+    def enabled?
+      Yast::UI.QueryWidget(Id(widget_id), :Enabled)
+    end
+
+    # Opens widget for modification
+    def enable
+      Yast::UI.ChangeWidget(Id(widget_id), :Enabled, true)
+    end
+
+    # Closes widget for modification
+    def disable
+      Yast::UI.ChangeWidget(Id(widget_id), :Enabled, false)
+    end
+
+  protected
+
+    # helper to check if event is invoked by this widget
+    def my_event?(widget, event)
+      return widget == event["ID"]
+    end
+
+    # shortcut from Yast namespace to avoid including whole namespace
+    # TODO: kill converts in CWM module, to avoid this workaround for funrefs
+    def fun_ref(*args)
+      Yast::FunRef.new(*args)
+    end
+
+  private
+
+    def init_method
+      fun_ref(method(:init), "void (string)")
+    end
+
+    def handle_method
+      fun_ref(method(:handle), "symbol (string, map)")
+    end
+
+    def store_method
+      fun_ref(method(:store), "void (string, map)")
+    end
+
+    def cleanup_method
+      fun_ref(method(:cleanup), "void (string)")
+    end
+
+    def validate_method
+      fun_ref(method(:validate), "boolean (string, map)")
+    end
+  end
+
+  # Represents custom widget, that have its UI content defined in method content.
+  # Useful mainly when specialized widget including more subwidget should be
+  # reusable at more places.
+  #
+  # @example custom widget child
+  #   class MyWidget < CWM::CustomWidget
+  #     def initialize
+  #       self.widget_id = "my_widget"
+  #     end
+  #
+  #     def content
+  #       HBox(
+  #         PushButton(Id(:reset), _("Reset")),
+  #         PushButton(Id(:undo), _("Undo"))
+  #       )
+  #     end
+  #
+  #     def handle(widget, event)
+  #       case event["ID"]
+  #       when :reset then ...
+  #       when :undo then ...
+  #       else ...
+  #       end
+  #     end
+  #   end
+  class CustomWidget < AbstractWidget
+    self.widget_type = :custom
+    # custom witget without content do not make sense
+    abstract_method :content
+
+    def description
+      {
+        "custom_widget" => content
+      }.merge(super)
```

Comment:
In multiple places I see `some_hash.merge(super)`, which makes it impossible to override the superclass values. Please reverse all the merges.